From 8af41c5d103a1d6c1452ebdd2d698dab080bba0f Mon Sep 17 00:00:00 2001
From: Gurumoorthy Santhakumar <gsanthak@qti.qualcomm.com>
Date: Tue, 13 May 2025 21:10:51 +0530
Subject: [PATCH] i2c: Add support for loading Serial Engine firmware

Introduce functionality to load peripheral-specific firmware required for initializing and enabling the I2C interface via the Serial Engine (SE).

This support is essential for platforms that rely on firmware-assisted initialization of I2C controllers, ensuring proper communication with I2C-based peripherals.

Change-Id: Ic8baa3f8ec937d6e488ef48628ba25b79585cba4
Signed-off-by: Gurumoorthy Santhakumar <gsanthak@qti.qualcomm.com>
---
 drivers/i2c/Kconfig    |  2 +-
 drivers/i2c/geni_i2c.c | 13 +++++++++++++
 2 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/drivers/i2c/Kconfig b/drivers/i2c/Kconfig
index cdae6825736..7754ec10284 100644
--- a/drivers/i2c/Kconfig
+++ b/drivers/i2c/Kconfig
@@ -640,7 +640,7 @@ config SYS_I2C_QUP
 
 config SYS_I2C_GENI
 	bool "Qualcomm Generic Interface (GENI) I2C controller"
-	depends on ARCH_SNAPDRAGON
+	depends on (ARCH_SNAPDRAGON || ARCH_IPQ)
 	help
 	  Support for the Qualcomm Generic Interface (GENI) I2C interface.
 	  The Generic Interface (GENI) is a firmware based Qualcomm Universal
diff --git a/drivers/i2c/geni_i2c.c b/drivers/i2c/geni_i2c.c
index eabf5c76c21..0b27a0e81b4 100644
--- a/drivers/i2c/geni_i2c.c
+++ b/drivers/i2c/geni_i2c.c
@@ -469,6 +469,8 @@ static int geni_i2c_probe(struct udevice *dev)
 	ofnode parent_node = ofnode_get_parent(dev_ofnode(dev));
 	struct geni_i2c_priv *geni = dev_get_priv(dev);
 	u32 proto, tx_depth, fifo_disable;
+	struct ofnode_phandle_args args;
+	struct udevice *config = NULL;
 	int ret;
 
 	geni->is_master_hub = dev_get_driver_data(dev) & GENI_I2C_IS_MASTER_HUB;
@@ -497,6 +499,17 @@ static int geni_i2c_probe(struct udevice *dev)
 
 	geni_i2c_enable_clocks(dev, geni);
 
+	ret = dev_read_phandle_with_args(dev, "qup-se-fw-load", NULL, 0, 0,
+					 &args);
+	if (!ret) {
+		ret = uclass_get_device_by_ofnode(UCLASS_NOP, args.node,
+						  &config);
+		if (ret)
+			dev_err(dev,
+				"Failed to write SE(I2C) Firmware: %d\n",
+				ret);
+	}
+
 	proto = readl(geni->base + GENI_FW_REVISION_RO);
 	proto &= FW_REV_PROTOCOL_MSK;
 	proto >>= FW_REV_PROTOCOL_SHFT;
-- 
2.34.1

